### PR TITLE
feat: Making bar graphs in Table viz from fixed-size divs instead of calculated gradients

### PR DESCRIPTION
**Pull request by rusackas**
### SUMMARY
The Table viz allows users to have little histogram/bar chart cells which can be gray, red when negative, or colorized via conditional formatting. Previously, these were implemented as the CSS background of the table cells, and a gradient was used to draw the bars, with gradient stops to set bar widths. This was problematic as people could not customized these with CSS, and it was also a potentially difficult to understand implementation from a maintenance standpoint.

In this revised approach, divs are conditionally added to table cells to act as these bars, and they're positioned behind the text content. They use a few small functions to calculate width, offset, and color for their display. They're also rendered with a `cell-bar` class and (when applicable) a `negative` or `positive` class to indicate if they're positive or negative, for further customization options.

image

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
No visual changes here - just representing what was changed. But, this is the "after" FWIW.

image

image

### TESTING INSTRUCTIONS

Test example tables, and throw different formatting options at the table. I think this covers all the use cases, but there's always a chance I missed something!

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API (CSS classes to customize with)
- [ ] Removes existing feature or API

### CODE REVIEW

**Review by michael-s-molina — APPROVED** [review]
LGTM. Left some non-blocking suggestions.

File: superset-frontend/plugins/plugin-chart-table/src/TableChart.tsx [hunk 1]
```diff
@@ -449,6 +491,15 @@ export default function TableChart<D extends DataRecord = DataRecord>(
           // render `Cell`. This saves some time for large tables.
           return (
             <StyledCell {...cellProps}>
+              {valueRange && (
+                <div
+                  className={cx(
+                    'cell-bar',
+                    value && value < 0 ? 'negative' : 'positive',
```
[the remark this answers is not in the document]
Comment (geido) [reply superset-frontend/plugins/plugin-chart-table/src/TableChart.tsx:498]: I think this is an important one!

File: superset-frontend/plugins/plugin-chart-table/src/TableChart.tsx [hunk 2]
```diff
@@ -404,16 +429,33 @@ export default function TableChart<D extends DataRecord = DataRecord>(
...
+            position: relative;
+            background: ${backgroundColor || undefined};
+          `;
+
+          const cellBarStyles = css`
+            position: absolute;
+            height: 100%;
+            display: block;
+            top: 0;
+            ${valueRange &&
```
[the remark this answers is not in the document]
Comment (rusackas) [reply superset-frontend/plugins/plugin-chart-table/src/TableChart.tsx:442]: This is actually just avoiding a bit of a typescript rabbit hole. The conditional isn't _really_ needed here, but if you take it out, it starts pulling the loose thread on a sweater...

[the remark this answers is not in the document]
Comment (rusackas) [reply superset-frontend/plugins/plugin-chart-table/src/TableChart.tsx:442]: I'll circle back to this if time allows.

File: superset-frontend/plugins/plugin-chart-table/src/TableChart.tsx [hunk 3]
```diff
@@ -80,44 +82,67 @@ function getSortTypeByDataType(dataType: GenericDataType): DefaultSortTypes {
...
   );
-  const perc2 = Math.round((Math.abs(value) / tot) * 100);
-  // The 0.01 to 0.001 is a workaround for what appears to be a
-  // CSS rendering bug on flat, transparent colors
-  return (
-    `linear-gradient(to right, rgba(0,0,0,0.01), rgba(0,0,0,0.001) ${perc1}%, ` +
-    `rgba(${r},0,0,0.2) ${perc1}%, rgba(${r},0,0,0.2) ${perc1 + perc2}%, ` +
-    `rgba(0,0,0,0.01) ${perc1 + perc2}%, rgba(0,0,0,0.001) 100%)`
-  );
+  return perc1;
```
[the remark this answers is not in the document]
Comment (rusackas) [reply superset-frontend/plugins/plugin-chart-table/src/TableChart.tsx:131]: It's a vestige of recycled code, but I liked the readability. You're right though, returning the value is straightforward enough. Done!

File: superset-frontend/plugins/plugin-chart-table/src/TableChart.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (rusackas) [reply superset-frontend/plugins/plugin-chart-table/src/TableChart.tsx:498]: Agreed. Comment added for posterity :D

**Review by ktmud — APPROVED** [review]

**Pull request merged**